# Deriving variables with for loops and if-statements

## Contains:
### Create price range column using user defined function for subset
### Create price range loc column using .loc for entire dataframe
### Create column busiest day and busiest days using for loop 
### Create column busiest period of day using for loop 

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#create text file path
path = r'C:\Users\krist\08.2023_InstacartBasketAnalysis'

In [3]:
#import orders_products_merged and check output
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02.Data', 'Prepared_Data', 'orders_products_merged.pkl'))
df_ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                   11.0         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices _exists  
0         Soda        77              7     9.0    both  
1         Soda        77              7     9.0    both  
2         Soda        77              7     9.0    both  
3         Soda        77              7     9.0    both  
4         Soda        77              7     9.0    both

In [4]:
#verify shape
df_ords_prods_merged.shape

(32404859, 15)

In [5]:
#create subset to use when practicing user defined function
df_sub = df_ords_prods_merged[:1000000]

In [6]:
#check subset shape
df_sub.shape

(1000000, 15)

In [7]:
#create function for price range label
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [8]:
#apply price_label function and put it in new column called price_range
df_sub['price_range'] = df_sub.apply(price_label, axis = 1)

C:\Users\krist\AppData\Local\Temp\ipykernel_11296\2340676141.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sub['price_range'] = df_sub.apply(price_label, axis = 1)


In [9]:
#check frequency of labels in price_range
df_sub['price_range'].value_counts()

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

### no High-range produts
### low-range = 243,550
### mid-range = 756,450

In [10]:
#find highest price in subset
df_sub['prices'].max()

14.8

### highest price in subset is 14.8

In [16]:
#use .loc() to address warning above
df_sub.loc[df_sub['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
#use .loc() to address warning above
df_sub.loc[(df_sub['prices'] <= 15) & (df_sub['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [15]:
#use .loc() to address warning above
df_sub.loc[df_sub['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
#check frequency of labels in price_range_loc column
df_sub['price_range_loc'].value_counts()

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

### frequencies are same as from price_label function

## Use loc() to apply filter to entire original dataframe 

In [18]:
#use loc() to apply filter and new label to original merged dataframe
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [20]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
#check frequencies for price range label 
df_ords_prods_merged['price_range_loc'].value_counts()

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

## Price range for products
### low-range 10,126,321
### mid-range 21,860,860
### high-range 417,678

In [22]:
#check frequencies of orders by day of week 
df_ords_prods_merged['order_day_of_week'].value_counts()

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

### busiest day 0 = Saturday
### least busy day 4 = Wednesday

In [29]:
result = []

In [30]:
#create for loop to determine what type of day products purchased on
for value in df_ords_prods_merged['order_day_of_week']:
  if value == 0:
    result.append('Busiest day')
  elif value == 4:
    result.append('Least busy')
  else:
    result.append('Regularly busy')

In [31]:
#check result list
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [32]:
#create new column where result list will go called 'busiest_day'
df_ords_prods_merged['busiest_day'] = result

In [33]:
#check frequency of busiest_day
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

### least busy count = day of week 4 count = 3,783,802
### busiest day count = day of week 0 count = 6,204,182

In [34]:
#checking new columns for end report
df_ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                   11.0         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices _exists    price_range_loc  \
0         Soda        77              7     9.0    both  Mid-range product   
1         Soda        77              7     9.0    both  Mid-range product   
2         Soda        77              7     9.0    both  Mid-range product   
3         Soda        77              7     9.0    both  Mid-range product   
4         Soda        77              7     9.0    both  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

## Create new column busiest_days with top two busiest days and bottom two busiest days 

### busiest days are 0 & 1, least busy are 3 & 4 

In [35]:
#create empty result_1 list for busiest_days 
result_1 = []

In [36]:
#two busiest days 0 & 1, two slowest days 3 & 4
#create for loop to determine what type of day products purchased on
for value in df_ords_prods_merged['order_day_of_week']:
  if (value == 0) | (value == 1):
    result_1.append('Busiest days')
  elif (value == 3) | (value == 4):
    result_1.append('Least busy days')
  else:
    result_1.append('Regularly busy')

In [38]:
#check subset of output from for loop
result_1[:100]

['Regularly busy',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy days',
 'Regularly busy',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Busiest day

In [40]:
#add new column to merged dataframe
df_ords_prods_merged['busiest_days'] = result_1

In [41]:
#check frequencies to compare to frequencies from order_day_of_week
df_ords_prods_merged['busiest_days'].value_counts()

Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: busiest_days, dtype: int64

### busiest days count matches sum of order day 0 and order day 1: 11,864,412
### least busy days count matches sum of order day 3 and order day 4: 7,624,336

## Create new column for busiest period of day with labels 'most orders', 'average orders', 'fewest orders'

In [46]:
#check frequencies to determine hours of day with highest, average, and lowest orders
df_ords_prods_merged['order_hour_of_day'].value_counts(sort = True)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [47]:
#assume frequencies over 2 million app will crash. hours 9,10,11,12,13,14,15,16,17 over 2 million
#assume frenquencies under 100 thousand considered not busy. hours 2,3,4,5 under 100 thousand
#assume all other frequencies considered average
#create results list for new column
result_2 = []

In [49]:
#create for loop for labels for hour of day
for value in df_ords_prods_merged['order_hour_of_day']:
    if (value >=2) and (value <=5):
        result_2.append('fewest orders')
    elif (value >=9) and (value <=17):
        result_2.append('most orders')
    else:
        result_2.append('average orders')

In [50]:
#check subset of result_2 for output
result_2[:100]

['average orders',
 'average orders',
 'most orders',
 'average orders',
 'most orders',
 'average orders',
 'most orders',
 'most orders',
 'most orders',
 'average orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'average orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'average orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'average orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'most orders',
 'average orders',
 'most orders',
 'most orders',
 'most orders

In [51]:
#add new column to dataframe and check frequencies of new column
df_ords_prods_merged['busiest_period_of_day'] = result_2
df_ords_prods_merged['busiest_period_of_day'].value_counts()

most orders       23205725
average orders     8937275
fewest orders       261859
Name: busiest_period_of_day, dtype: int64

### value of most orders matches sum of hours 9-17: 23,205,725
### value of fewest orders matches sum of hours 2-5: 261,859
### value of average orders matches sum of remaining hours: 8,937,275

In [52]:
#check shape of dataframe after derivations
df_ords_prods_merged.shape

(32404859, 19)

### dataframe after column derivations has 32,404,859 rows 19 columns

## Export to pkl

In [53]:
#export dataframe to pkl
df_ords_prods_merged.to_pickle(os.path.join(path, '02.Data', 'Prepared_Data', 'orders_products_merged_4.7.pkl'))